triton-vm/src/table/constraint_circuit.rs

```diff
@@ -1060,6 +875,182 @@ impl<II: InputIndicator> ConstraintCircuitMonad<II> {
+            let new_col_idx = match chosen_node_is_base_col {
+                true => num_base_cols + base_constraints.len(),
+                false => num_ext_cols + ext_constraints.len(),
+            };
+            let new_input_indicator = match chosen_node_is_base_col {
+                true => II::base_table_input(new_col_idx),
+                false => II::ext_table_input(new_col_idx),
+            };
+            let new_variable = builder.input(new_input_indicator);
+            let new_circuit = new_variable.circuit.clone();
+
+            // Substitute the chosen circuit with the new variable.
+            builder.substitute(chosen_node_id, new_circuit.clone());
+
+            // Create new constraint and put it into the appropriate return vector.
+            let new_constraint = new_variable - chosen_node;
+            match chosen_node_is_base_col {
+                true => base_constraints.push(new_constraint),
+                false => ext_constraints.push(new_constraint),
+            }
+
+            // Treat roots of the multicircuit explicitly.
+            for circuit in multicircuit.iter_mut() {
+                if circuit.circuit.borrow().id == chosen_node_id {
+                    circuit.circuit = new_circuit.clone();
+                }
+            }
+        }
+
+        (base_constraints, ext_constraints)
+    }
+
+    /// Heuristically pick a node from the given multicircuit that is to be substituted with a new
+    /// variable. The ID of the chosen node is returned.
+    fn pick_node_to_substitute(
+        multicircuit: &[ConstraintCircuitMonad<II>],
+        target_degree: Degree,
+    ) -> usize {
+        if multicircuit.is_empty() {
+            panic!("Multicircuit must be non-empty in order to pick a node from it.");
+        }
+
+        let multicircuit = multicircuit
+            .iter()
+            .map(|c| c.clone().consume())
+            .collect_vec();
+        let all_nodes = Self::all_nodes_in_multicircuit(&multicircuit);
+        let all_nodes: HashSet<_> = HashSet::from_iter(all_nodes.into_iter());
+
+        // Only nodes with degree > target_degree need changing.
+        let high_degree_nodes = all_nodes
+            .into_iter()
+            .filter(|node| node.degree() > target_degree)
+            .collect_vec();
+
+        // Collect all the nodes where some substitution is necessary.
+        // Substituting a node of degree 1 is both pointless and can lead to infinite iteration.
+        let low_degree_nodes = Self::all_nodes_in_multicircuit(&high_degree_nodes)
+            .into_iter()
+            .filter(|node| node.degree() <= target_degree)
+            .filter(|node| node.degree() > 1)
+            .collect_vec();
+
+        // If the resulting list is empty, there is no way forward. Stop – panic time!
+        assert!(
+            !low_degree_nodes.is_empty(),
+            "Could not lower degree of circuit to target degree. This is a bug."
+        );
```

Comment:
I don't think it's ever possible that all nodes are of high degree, so long as “high” means greater than 1. In particular, a leaf in the multicircuit has degree at most 1, which is not high by assumption. Therefore, a high degree node must be an internal node. There exist leafs in the tree making up this internal node.

All that said, nodes of degree smaller than or equal to 1 are explicitly excluded as substitution candidates, because such a substitution would not change the degree, giving rise to termination problems. That's the original reason for having put above assertion.